### PR TITLE
Improve subscribe loop performance, rename CoreOperator

### PR DESCRIPTION
**Pull request by bsideup**
This PR removes expensive `instanceof` checks from the tailrec loop without exposing it to the public API (as seen in https://github.com/reactor/reactor-core/compare/master...tail_call_subscribe_3?expand=1 )

# Benchmarks
Shakespeare:
```
───────┬────────────────────────────────────────────────────────────────────────────────────────────
       │ File: build/reports/jmh/result.text
───────┼────────────────────────────────────────────────────────────────────────────────────────────
   1   │ Benchmark                                              Mode  Cnt    Score   Error  Units
   2   │ ShakespearePlaysScrabbleOpt.measureThroughput          avgt   25   34,898 ± 0,471  ms/op
   3   │ ShakespearePlaysScrabbleParallelOpt.measureThroughput  avgt   25   36,944 ± 0,435  ms/op
   4   │ ShakespearePlaysScrabbleRegular.measureThroughput      avgt   25  118,314 ± 0,910  ms/op
───────┴────────────────────────────────────────────────────────────────────────────────────────────
───────┬────────────────────────────────────────────────────────────────────────────────────────────
       │ File: build/reports/jmhBaseline/result.text
───────┼────────────────────────────────────────────────────────────────────────────────────────────
   1   │ Benchmark                                              Mode  Cnt    Score   Error  Units
   2   │ ShakespearePlaysScrabbleOpt.measureThroughput          avgt   25   38,821 ± 2,083  ms/op
   3   │ ShakespearePlaysScrabbleParallelOpt.measureThroughput  avgt   25   41,844 ± 2,401  ms/op
   4   │ ShakespearePlaysScrabbleRegular.measureThroughput      avgt   25  125,757 ± 5,908  ms/op
───────┴────────────────────────────────────────────────────────────────────────────────────────────
```

TailCall:
```
───────┬────────────────────────────────────────────────────────────────────────────────────────────
       │ File: build/reports/jmh/result.text
───────┼────────────────────────────────────────────────────────────────────────────────────────────
   1   │ Benchmark               (operatorsCount)  Mode  Cnt      Score     Error  Units
   2   │ TailCallBenchmark.with                 5  avgt    5    381,104 ±   1,751  ns/op
   3   │ TailCallBenchmark.with                50  avgt    5   6837,949 ± 123,034  ns/op
   4   │ TailCallBenchmark.with               100  avgt    5  14004,323 ± 409,916  ns/op
───────┴────────────────────────────────────────────────────────────────────────────────────────────
───────┬────────────────────────────────────────────────────────────────────────────────────────────
       │ File: build/reports/jmhBaseline/result.text
───────┼────────────────────────────────────────────────────────────────────────────────────────────
   1   │ Benchmark               (operatorsCount)  Mode  Cnt      Score     Error  Units
   2   │ TailCallBenchmark.with                 5  avgt    5    405,893 ±  12,211  ns/op
   3   │ TailCallBenchmark.with                50  avgt    5   6958,641 ±  85,940  ns/op
   4   │ TailCallBenchmark.with               100  avgt    5  14184,554 ± 343,604  ns/op
───────┴────────────────────────────────────────────────────────────────────────────────────────────
```

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/FluxReplay.java [hunk 1]
```diff
@@ -1091,6 +1095,10 @@ public void subscribe(CoreSubscriber<? super T> actual) {
 			c.buffer.replay(inner);
 
 			if (expired) {
+				if (coreOperator == null) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxReplay.java]: is this change really needed?

Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/FluxReplay.java]: yes, part of the contract

File: reactor-core/src/main/java/reactor/core/publisher/Operators.java [hunk 2]
```diff
@@ -1287,12 +1291,16 @@ public void subscribe(Subscriber<? super T> s) {
 
 		@Override
 		public CoreSubscriber<? super T> subscribeOrReturn(CoreSubscriber<? super T> actual) {
+			if (coreOperator == null) {
```
[the remark this answers is not in the document]
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java]: Could you please clarify? `CorePublisherAdapter` can still appear in between of two `CoreOperator`s

[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Operators.java]: I thought for a minute that since all subscribe methods were implemented for this class, `subscribeOrReturn` wouldn't be called. Carry-on 🙂

**Comment by simonbasle** [comment]
in the latest state, there's several classes which only change is the newline at end of file. can you make sure to revert these so that the PR doesn't unnecessarily appear in their history @bsideup ?
